auth/auth-server/src/server/handle_external_match/gas_sponsorship/refund_calculation.rs

```diff
@@ -0,0 +1,125 @@
+//! Logic for calculating refund info for a sponsored match
+
+use alloy_primitives::U256 as AlloyU256;
+use bigdecimal::{BigDecimal, FromPrimitive};
+use renegade_api::http::external_match::{ApiExternalMatchResult, ApiExternalQuote};
+use renegade_circuit_types::order::OrderSide;
+use renegade_common::types::token::Token;
+use renegade_constants::NATIVE_ASSET_ADDRESS;
+
+use crate::{
+    error::AuthServerError,
+    server::{
+        helpers::{ethers_u256_to_alloy_u256, get_nominal_buy_token_price},
+        Server,
+    },
+};
+
+// -------------
+// | Constants |
+// -------------
+
+/// The number of Wei in 1 ETH, as an `AlloyU256`.
+/// Concretely, this is 10^18
+const ALLOY_WEI_IN_ETHER: AlloyU256 =
+    AlloyU256::from_limbs([1_000_000_000_000_000_000_u64, 0, 0, 0]);
+
+// ---------------
+// | Server Impl |
+// ---------------
+
+impl Server {
+    /// Get the amount to refund for a given match result
+    pub async fn get_refund_amount(
+        &self,
+        match_result: &ApiExternalMatchResult,
+        refund_native_eth: bool,
+    ) -> Result<AlloyU256, AuthServerError> {
+        let conversion_rate =
+            self.maybe_fetch_conversion_rate(match_result, refund_native_eth).await?;
+
+        let estimated_gas_cost = ethers_u256_to_alloy_u256(self.get_gas_cost_estimate().await);
+
+        let refund_amount = if let Some(conversion_rate) = conversion_rate {
+            (estimated_gas_cost * conversion_rate) / ALLOY_WEI_IN_ETHER
+        } else {
+            estimated_gas_cost
+        };
+
+        Ok(refund_amount)
+    }
+
+    /// Fetch the conversion rate from ETH to the buy-side token in the trade
+    /// from the price reporter, if necessary.
+    /// The conversion rate is in terms of nominal units of TOKEN per whole ETH.
+    #[allow(clippy::unused_async)]
+    async fn maybe_fetch_conversion_rate(
```

Comment:
this method was directly copied over from `gas_sponsorship/mod.rs`, the other 2 in this file are new/updated and should be reviewed